config/config.go

```diff
@@ -29,186 +29,248 @@ import (
+	l.lock.RLock()
+	defer l.lock.RUnlock()
 
-	dirs := []string{".", _configRoot}
-	for _, dir := range dirs {
+	var files []string
+	for _, dir := range l.dirs {
 		for _, baseFile := range fileSet {
-			files = append(files, fmt.Sprintf("%s/%s.yaml", dir, baseFile))
+			files = append(files, filepath.Join(dir, fmt.Sprintf("%s.yaml", baseFile)))
 		}
 	}
+
 	return files
 }
 
-func baseFiles() []string {
-	env := Environment()
-	return []string{_baseFile, env, _secretsFile}
+func (l *Loader) baseFiles() []string {
+	return []string{_baseFile, l.Environment(), _secretsFile}
 }
 
-func getResolver() FileResolver {
-	paths := []string{}
-	configDir := Path()
-	if configDir != "" {
-		paths = []string{configDir}
+func (l *Loader) getResolver() FileResolver {
+	if dir := l.Path(); dir != "" {
+		NewRelativeResolver(dir)
 	}
-	return NewRelativeResolver(paths...)
+
+	return NewRelativeResolver()
 }
 
 // YamlProvider returns function to create Yaml based configuration provider
-func YamlProvider() ProviderFunc {
+func (l *Loader) YamlProvider() ProviderFunc {
 	return func() (Provider, error) {
-		return NewYAMLProviderFromFiles(false, getResolver(), getConfigFiles(_configFiles...)...), nil
+		return NewYAMLProviderFromFiles(false, l.getResolver(), l.joinFilepaths(l.getFiles()...)...), nil
 	}
 }
 
 // Environment returns current environment setup for the service
-func Environment() string {
-	env := os.Getenv(EnvironmentKey())
-	if env == "" {
-		env = _devEnv
+func (l *Loader) Environment() string {
+	if env, ok := l.lookUp(l.EnvironmentKey()); ok {
+		return env
 	}
-	return env
+
+	return _devEnv
 }
 
 // Path returns path to the yaml configurations
-func Path() string {
-	configPath := os.Getenv(EnvironmentPrefix() + _configDir)
-	if configPath == "" {
-		configPath = _configRoot
+func (l *Loader) Path() string {
+	if path, ok := l.lookUp(l.EnvironmentPrefix() + _configDir); ok {
+		return path
 	}
-	return configPath
+
+	return _configRoot
 }
 
-// SetConfigFiles overrides the set of available config files
-// for the service
-func SetConfigFiles(files ...string) {
-	_configFiles = files
+// SetConfigFiles overrides the set of available config files for the service.
+func (l *Loader) SetConfigFiles(files ...string) {
+	l.lock.Lock()
+
+	l.configFiles = files
+
+	l.lock.Unlock()
+}
+
+func (l *Loader) getFiles() []string {
+	l.lock.RLock()
+	l.lock.RUnlock()
+
+	res := make([]string, len(l.configFiles))
+	copy(res, l.configFiles)
+	return res
+}
+
+// SetDirs overrides the set of dirs to load config files from.
+func (l *Loader) SetDirs(dirs ...string) {
+	l.lock.Lock()
+
+	l.dirs = dirs
+
+	l.lock.Unlock()
 }
 
-// SetEnvironmentPrefix sets environment prefix for the application
-func SetEnvironmentPrefix(envPrefix string) {
-	_envPrefix = envPrefix
+// SetEnvironmentPrefix sets environment prefix for the application.
+func (l *Loader) SetEnvironmentPrefix(envPrefix string) {
+	l.lock.Lock()
+
+	l.envPrefix = envPrefix
+
+	l.lock.Unlock()
```

Comment:
Well, I was doing for performance, but consistency is probably going to be safer :)